internal/gcs/container.go

```diff
@@ -225,20 +237,63 @@ func (c *Container) Terminate(ctx context.Context) (err error) {
 // Wait waits for the container to terminate (or Close to be called, or the
 // guest connection to terminate).
 func (c *Container) Wait() error {
-	select {
-	case <-c.notifyCh:
-		return nil
-	case <-c.closeCh:
+	<-c.closeCh
+
+	// c.Close() will set `c.gc = nil`
+	if c.gc == nil {
 		return errors.New("container closed")
 	}
+
+	return nil
+}
+
+// Notifications returns a list of notifications (including shutdown) about the container
+func (c *Container) Notifications() (*queue.MessageQueue, error) {
+	return c.notifications, nil
 }
 
-func (c *Container) waitBackground() {
-	ctx, span := trace.StartSpan(context.Background(), "gcs::Container::waitBackground")
+// should only be called once; will wait for container exit and notify other
+// `Wait()` calls, as well as update notification buffer
+func (c *Container) waitBackground(ctx context.Context) {
+	var err error
+	ctx, span := trace.StartSpan(ctx, "gcs::Container::waitBackground")
 	defer span.End()
-	span.AddAttributes(trace.StringAttribute("cid", c.id))
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.ContainerID, c.id))
+
+	// read in notifications while waiting for container to exit
+	loop := true
+	for loop {
+		select {
+		case <-c.closeCh:
+			loop = false
+			// c.Close() will set `c.gc = nil`
+			if c.gc == nil {
+				err = errors.New("container closed")
+			}
+		case <-ctx.Done():
+			loop = false
+		case ntf, ok := <-c.notifyCh:
+			if !ok {
+				// channel was closed
+				loop = false
+			}
+			// enqueue the notification; will block until write succeeds
+			err = c.notifications.Write(ntf)
+			if err != nil {
+				log.G(ctx).WithError(err).
+					WithField("notification", ntf.String()).
+					Error("could not enqueue notification")
+			}
+		}
+
+	}
+
+	c.closeOnce.Do(func() {
+		// `notifyCh` is closed by `GuestConnection.notify`, so do not close here
+		close(c.closeCh)
+		c.notifications.Close()
```

Comment:
Why aren't we calling wait anymore? This would prevent a call to `c.Close` from running anything.

Reply:
Originally `c.Close` did nothing, and looking at [`hcs.System`](https://github.com/microsoft/hcsshim/blob/master/internal/hcs/system.go#L271-L274) and [`jobcontainers.JobContainer`](https://github.com/microsoft/hcsshim/blob/master/internal/jobcontainers/jobcontainer.go#L505-L508), they also closed their resources at the end of their `waitBackground` using the same `Once` as their `Close`, so I adopted a similar approach, where `Close` ends the connection, but since it does not own any resources, there is no backing container to close.
I can change it back if `Container` is intended to be used in a different way